Added helper script and cron definition to run it, to keep log sizes within limits even for runaway build jobs.

andreas-kupries opened the pull request:
Ref https://trello.com/c/oUZP2WTV/613-jenkins-healthcheck-and-more-documentation

andreas-kupries committed:
Added helper script and cron definition to run it, to keep log sizes within limits even for runaway build jobs.

mook-as commented on provisioners/file/home/ec2-user/bin/limit-logs:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+cd /data/jenkins/jobs
+sudo find . -type f -name '*log' -size +15M \
+    -exec truncate --size 15M {} \;
```
Supposedly `execdir` is safer, though I never quite understood why.

andreas-kupries replied:
Per find manpage avoids races in path resolution.
Changed with commit 5f4ac1ddbc5349bd9d607afb808001070b44de43

mook-as commented on provisioners/file/home/ec2-user/bin/limit-logs:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+cd /data/jenkins/jobs
+sudo find . -type f -name '*log' -size +15M \
```
You don't need the `cd` if you just `find /data/jenkins/jobs -type …`
Also, does that actually need sudo? Shouldn't the logs be owned by jenkins?

andreas-kupries replied:
Actually I ran into permission errors without the sudo. Lets see if I still have them in the terminal buffer.

andreas-kupries replied:
```
ec2-user@ip-172-31-37-206:/data/jenkins/jobs> limit-logs 
truncate: cannot open './scf-master-sles/builds/56/110.log' for writing: Permission denied
truncate: cannot open './scf-master-sles/builds/56/log' for writing: Permission denied
truncate: cannot open './scf-develop/builds/267/log' for writing: Permission denied
truncate: cannot open './scf-develop/builds/267/95.log' for writing: Permission denied
truncate: cannot open './scf/branches/PR-1436/builds/1/89.log' for writing: Permission denied
truncate: cannot open './scf/branches/PR-1436/builds/1/log' for writing: Permission denied
```

Actually these might have been the logs you truncated before, they were all of size 101M.
And if you did that via root or sudo, then they may have become root owned.

Checking ownership of the logs now ...

andreas-kupries replied:
Ah! Logs are owned by jenkins. I am however user `ec2-user`.
Lets see if I can shift the cron job over to user `jenkins`.

andreas-kupries replied:
Changed with commit 5f4ac1ddbc5349bd9d607afb808001070b44de43

jandubois commented on provisioners/file/home/ec2-user/bin/limit-logs:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+cd /data/jenkins/jobs
+sudo find . -type f -name '*log' -size +15M \
```
Why is this running as `ec2-user` and then needs sudo?

Wouldn't it be better to run this as a root cron job?

jandubois replied:
This is just personal preference, but I prefer

```
find /dir ...
```

over

```
cd /dir
find . ...
```

It doesn't matter for a trivial script like this, but the former version remains valid even when the cwd is changed elsewhere.

andreas-kupries replied:
Sure, can make that change. Re the sudo, see Mark's comment and my answers.

jandubois replied:
Github never showed me Mark's comments when I added mine; otherwise I wouldn't have bothered 😄

andreas-kupries replied:
Changed with commit 5f4ac1ddbc5349bd9d607afb808001070b44de43

andreas-kupries committed:
Address comments
- Move cron job to jenkins user, drop sudo
- find exactly from directory
- execdir to avoid path resolution race

mook-as commented on provisioners/file/var/spool/cron/tabs/jenkins:
```diff
@@ -0,0 +1,4 @@
+# DO NOT EDIT THIS FILE - edit the master and reinstall.
+# (C installed on Thu Apr  5 21:42:22 2018)
```
Technically this _is_ the master file, since you're not using `crontab -e` here.

andreas-kupries replied:
Sure, will fix.

andreas-kupries replied:
Done with commit 02aa7bfb831474f63a5c0d24456a99b47906f446

mook-as commented on provisioners/file/var/spool/cron/tabs/jenkins:
```diff
@@ -0,0 +1,4 @@
+# DO NOT EDIT THIS FILE - edit the master and reinstall.
+# (C installed on Thu Apr  5 21:42:22 2018)
+# (Cronie version 4.2)
+43 * * * * /bin/bash /home/ec2-user/bin/limit-logs
```
This really shouldn't live in `ec2-user`'s home now that they don't run it anymore…

andreas-kupries replied:
What other nice location do we have for it ?

mook-as replied:
Some place in `/data/jenkins` I guess? `/usr/local/bin`?

andreas-kupries replied:
:eyes: now ...

andreas-kupries replied:
I dislike `/data/jenkins` because I feel that I cannot be sure that I will not be trampled upon regardless of what name I use for directory and script. `/usr/local/bin` looks quite empty and not owned by jenkins. I will move there.

andreas-kupries replied:
Done with commit 1ea634ba15df62cdf187eec83a0a08e792d1cbf6

andreas-kupries committed:
Remove `crontab -e`-generated comments, not true.

andreas-kupries committed:
Move helper script to /usr/local/bin and adjust the crontab.
